examples/consistency_models/train_consistency_distillation.py

```diff
@@ -0,0 +1,701 @@
+            dirs = [d for d in dirs if d.startswith("checkpoint")]
+            dirs = sorted(dirs, key=lambda x: int(x.split("-")[1]))
+            path = dirs[-1] if len(dirs) > 0 else None
+
+        if path is None:
+            accelerator.print(
+                f"Checkpoint '{args.resume_from_checkpoint}' does not exist. Starting a new training run."
+            )
+            args.resume_from_checkpoint = None
+        else:
+            accelerator.print(f"Resuming from checkpoint {path}")
+            accelerator.load_state(os.path.join(args.output_dir, path))
+            global_step = int(path.split("-")[1])
+
+            initial_global_step = global_step * args.gradient_accumulation_steps
+            first_epoch = global_step // num_update_steps_per_epoch
+    else:
+        initial_global_step = 0
+
+    progress_bar = tqdm(
+        range(0, max_train_steps),
+        initial=initial_global_step,
+        desc="Steps",
+        # Only show the progress bar once on each machine.
+        disable=not accelerator.is_local_main_process,
+    )
+
+
+
+    # Train!
+    for epoch in range(first_epoch, args.num_epochs):
+        model.train()
+        for step, batch in enumerate(train_dataloader):
+            clean_images = batch["input"]
+            labels = batch["labels"]
+            # Sample noise that we'll add to the images
+            noise = torch.randn(clean_images.shape).to(clean_images.device)
+            # Sample a random timestep for each image, TODO - allow different timesteps in a batch
+            index = torch.randint(
+                0, noise_scheduler.config.num_train_timesteps-1,  (1,), device=clean_images.device
+            ).long()
+            timestep = timesteps[index]
+            timestep_prev = timestep + 1
+            # TODO, we should have an add noise in the scheduler maybe?
+            noised_image = clean_images + noise*append_dims(timestep, clean_images.ndim)
+            scaled_timesteps = noise_scheduler.scale_timestep(timestep)
+            scaled_timesteps_prev = noise_scheduler.scale_timestep(timestep_prev) 
+            target_model_ema.copy_to(target_model.parameters())
+
+            with accelerator.accumulate(model):
+                # Predict the noise residual
+                model_output = model(noised_image, scaled_timesteps, class_labels=labels).sample
+                distiller = noise_scheduler.step(
+                    model_output, timestep, noised_image, use_noise=False
+                ).prev_sample
+
+                # Heun Solver to get previous timestep image using teacher model
+                # TODO - make this cleaner
+                samples = noised_image
+                x = samples
+                model_output = teacher_model(x, scaled_timesteps, class_labels=labels).sample
+                teacher_denoiser = noise_scheduler.step(
+                    model_output, timestep, x, use_noise=False
+                ).prev_sample
```

Comment:
Similarly to https://github.com/dg845/diffusers/pull/2/files#r1233164339, I believe this should use the teacher model's scheduler rather than `noise_scheduler`, which is the student consistency model's scheduler.

Alternatively, maybe we should not call `teacher_scheduler.step` and instead always take a Heun step to get the teacher `denoised_image` (as defined on Line 598 below). Not sure which one makes more sense; I believe that the EDM models used as the teacher models for the consistency distillation examples in the original code all happen to be trained with the Heun sampler as well. Section 4 in the [paper](https://arxiv.org/pdf/2303.01469.pdf) leads me to believe that any sampler can be used, not just the sampler the teacher model was trained on. @williamberman thoughts?

Reply:
Hey @dg845 , I think the teacher model would _always_ use the Heun/Euler solver of the paper, since it is only used as a score function to provide the score `s_phi(x, t)`, and not as a model with its own independent sampler. My understanding is that it is similar to the classifiers used in classifier guidance, and we don't really care about how it was trained.